### PR TITLE
Also source questions and answers from the Code4SA / PMG API

### PR DESCRIPTION
**Pull request by mhl**
Code4SA have added questions and answers to the PMG API, and this has
much better data than our scrapers have found since mid-2014.  This
commit adds to the scraper an extra stage, which creates Question and
Answer objects from data from the PMG API.

Fixes mysociety/pombola#1797

### CODE REVIEW

File: za_hansard/management/commands/za_hansard_q_and_a_scraper.py [hunk 1]
```diff
@@ -33,6 +38,46 @@
...
+    q_list = [
+        (text_similarity(
+            tidy_for_matching(question.question),
+            tidy_for_matching(target)),
+         question)
+        for question in questions_qs
+    ]
+    if not q_list:
+        raise Question.DoesNotExist()
+    ratio, question = min(q_list)
```
Comment (struan) [inline za_hansard/management/commands/za_hansard_q_and_a_scraper.py]: It is not clear to me why min works here given the second value in the tuple is a question object.

File: za_hansard/management/commands/za_hansard_q_and_a_scraper.py [hunk 2]
```diff
@@ -239,6 +302,148 @@ def scrape_answers(self, start_url_a, *args, **options):
...
+        existing_kwargs.update(number_q_kwargs)
+        if not number_found:
+            # We won't be able to accurately tell whether a question
+            # already exists if we don't have one of these number, so
+            # ignore the question and answer completely in that
+            # case. (This is a rare occurence.)
+            print "Skipping {0} because no number was found".format(data['url'])
+            return
+        try:
+            question = find_matching_question(
```
Comment (struan) [inline za_hansard/management/commands/za_hansard_q_and_a_scraper.py]: I don't really understand the data well enough to know why we need to do this. If we are matching using a number then why do we need the text comparison code?

Comment (mhl) [inline za_hansard/management/commands/za_hansard_q_and_a_scraper.py]: There are some cases where there are multiple questions with the same number within a year already in the database. I think this is almost always a result of a parsing error, in which case picking the one with the most similar text seemed like the most obvious way of picking the right one, and not the one with a misparsed number. I guess I should add a comment saying more or less that, if that makes sense!

File: za_hansard/management/commands/za_hansard_q_and_a_scraper.py (hunk 1 above)
Comment (mhl) [inline za_hansard/management/commands/za_hansard_q_and_a_scraper.py]: It's basically always the first element of the tuple (the similarity ratio) that dictates the order between two tuples - if it comes down to comparing the question objects, it'll be arbitrary (as per discussion in #ms, it'll be based on the Python object ID), but if the similarity is *exactly* the same, we probably don't care anyway.

File: za_hansard/management/commands/za_hansard_q_and_a_scraper.py (hunk 2 above)
Comment (mhl) [inline za_hansard/management/commands/za_hansard_q_and_a_scraper.py]: As I mentioned in stand-up, this was completely wrong-headed - I'm not sure how I got confused about this. There's a unique constraint that means that there can't be more than one question or answer with the same `written_answer`, `house` and `year`. So, I've removed all of that code.

File: za_hansard/tests/pmg_q_and_a_tests.py [hunk 3]
```diff
@@ -241,3 +241,67 @@ def api_one_question_and_answer(url):
...
+                return
+            elif url == 'https://api.pmg.org.za/member/':
+                return
+            elif url == 'http://api.pmg.org.za/minister/2/questions/':
+                yield EXAMPLE_QUESTION
+            else:
+                raise Exception("Unfaked URL '{0}'".format(url))
+        fake_all_from_api.side_effect = api_one_question_and_answer
+        # Create an existing question and answer with the right year
+        # and written_number, but no corresponding existing question.
```
Comment (struan) [inline za_hansard/tests/pmg_q_and_a_tests.py]: I think this comment is misleading in the it's only creating an answer?

**Comment by struan** [comment]
Misleading comment aside this looks good.

File: za_hansard/tests/pmg_q_and_a_tests.py (hunk 3 above)
Comment (mhl) [inline za_hansard/tests/pmg_q_and_a_tests.py]: Oops, yes - I'll fix that and force push to update.

**Pull request merged**